### DISCUSSION TITLE
How to block Ip based on modsec rule.

### DISCUSSION THREAD

**Discussion by uhlhosting** [fail2ban/fail2ban#3489] [Q&A] [answered]
Hi,

So far I have the following:

In the bellow logs, I use grep against **148.72.232.65** one IP blocked by the following modsec rule:

ip-blacklist.conf contains
```bash
SecRule REMOTE_ADDR ".*" "chain,phase:1,t:none,log,block,id:'2200000',msg:'SLR: Client IP in Blacklist.',tag:'AUTOMATION/MALICIOUS',setvar:'tx.ip_blacklist=/%{matched_var}/'"
        SecRule TX:IP_BLACKLIST "@pmFromFile ip_blacklist.txt"
```

ip_blacklist.txt contains
```bash
/148.72.232.65/
````

These are the logs generated by the rule:

```bash
[root@cp httpd]# grep -Rn 148.72.232.65
modsec_debug.log:212:[02/Apr/2023:07:24:37.728826 +0200] [musashi.ninja/sid#55cdb7739580][rid#7f1d80016590][/xmlrpc.php][1] Access denied with code 403 (phase 1). Matched phrase "/148.72.232.65/" at TX:ip_blacklist. [file "/etc/httpd/modsecurity.d/activated_rules/ip-blacklist.conf"] [line "2"] [id "2200000"] [msg "SLR: Client IP in Blacklist."] [tag "AUTOMATION/MALICIOUS"]
modsec_debug.log:907:[02/Apr/2023:16:53:37.223071 +0200] [xn--vakuumwrmedmmung-1nbe.ch/sid#55cdb76b4978][rid#7f1d9c055c10][/xmlrpc.php][1] Access denied with code 403 (phase 1). Matched phrase "/148.72.232.65/" at TX:ip_blacklist. [file "/etc/httpd/modsecurity.d/activated_rules/ip-blacklist.conf"] [line "2"] [id "2200000"] [msg "SLR: Client IP in Blacklist."] [tag "AUTOMATION/MALICIOUS"]
modsec_debug.log:910:[02/Apr/2023:16:53:54.169344 +0200] [xn--vakuumwrmedmmung-1nbe.ch/sid#55cdb76b4978][rid#7f1d9c1ec8c0][/xmlrpc.php][1] Access denied with code 403 (phase 1). Matched phrase "/148.72.232.65/" at TX:ip_blacklist. [file "/etc/httpd/modsecurity.d/activated_rules/ip-blacklist.conf"] [line "2"] [id "2200000"] [msg "SLR: Client IP in Blacklist."] [tag "AUTOMATION/MALICIOUS"]
modsec_debug.log:1032:[02/Apr/2023:18:18:16.722269 +0200] [xn--vakuumwrmedmmung-1nbe.ch/sid#55cdb76b4978][rid#7f1db0095540][/xmlrpc.php][1] Access denied with code 403 (phase 1). Matched phrase "/148.72.232.65/" at TX:ip_blacklist. [file "/etc/httpd/modsecurity.d/activated_rules/ip-blacklist.conf"] [line "2"] [id "2200000"] [msg "SLR: Client IP in Blacklist."] [tag "AUTOMATION/MALICIOUS"]
modsec_debug.log:1199:[02/Apr/2023:21:48:13.663397 +0200] [xn--vakuumwrmedmmung-1nbe.ch/sid#55cdb76b4978][rid#7f1dd8004c20][/xmlrpc.php][1] Access denied with code 403 (phase 1). Matched phrase "/148.72.232.65/" at TX:ip_blacklist. [file "/etc/httpd/modsecurity.d/activated_rules/ip-blacklist.conf"] [line "2"] [id "2200000"] [msg "SLR: Client IP in Blacklist."] [tag "AUTOMATION/MALICIOUS"]
modsec_debug.log:1209:[02/Apr/2023:21:57:43.765440 +0200] [www.autojanser.ch/sid#55cdb7768f18][rid#7f1d8402ce40][/xmlrpc.php][1] Access denied with code 403 (phase 1). Matched phrase "148.72.232.65/" at TX:ip_blacklist. [file "/etc/httpd/modsecurity.d/activated_rules/ip-blacklist.conf"] [line "2"] [id "2200000"] [msg "SLR: Client IP in Blacklist."] [tag "AUTOMATION/MALICIOUS"]
modsec_audit.log:470:musashi.ninja 148.72.232.65 - - [02/Apr/2023:07:24:37.729213 +0200] "POST /xmlrpc.php HTTP/1.1" 403 199 "-" "-" ZCkRlV8Kg2J8XDdQFir4ugAAAAo "-" /apache/20230402/20230402-0724/20230402-072437-ZCkRlV8Kg2J8XDdQFir4ugAAAAo 0 2906 md5:e7d094f77cf34b384bb2d9555229ce56 
modsec_audit.log:3222:xn--vakuumwrmedmmung-1nbe.ch 148.72.232.65 - - [02/Apr/2023:16:53:37.223398 +0200] "POST /xmlrpc.php HTTP/1.1" 403 199 "-" "-" ZCmW8a3__A1h-X3sfW8Y5wAAARI "-" /apache/20230402/20230402-1653/20230402-165337-ZCmW8a3__A1h-X3sfW8Y5wAAARI 0 2935 md5:89e0584d3eddee8984321dbd6afd4b74 
modsec_audit.log:3226:xn--vakuumwrmedmmung-1nbe.ch 148.72.232.65 - - [02/Apr/2023:16:53:54.169648 +0200] "POST /xmlrpc.php HTTP/1.1" 403 199 "-" "-" ZCmXAmZ89A6jgsNgrBjnKwAAAEo "-" /apache/20230402/20230402-1653/20230402-165354-ZCmXAmZ89A6jgsNgrBjnKwAAAEo 0 2936 md5:4cc0857689f533f9a382f02c17ecca62 
modsec_audit.log:3678:xn--vakuumwrmedmmung-1nbe.ch 148.72.232.65 - - [02/Apr/2023:18:18:16.722570 +0200] "POST /xmlrpc.php HTTP/1.1" 403 199 "-" "-" ZCmqyF8Kg2J8XDdQFioPKwAAAAM "-" /apache/20230402/20230402-1818/20230402-181816-ZCmqyF8Kg2J8XDdQFioPKwAAAAM 0 2886 md5:98bf8ee6641cbb9a06a0ac60f4b14a6c 
modsec_audit.log:4436:xn--vakuumwrmedmmung-1nbe.ch 148.72.232.65 - - [02/Apr/2023:21:48:13.663741 +0200] "POST /xmlrpc.php HTTP/1.1" 403 199 "-" "-" ZCnb_YjvQ81K0fl0xJUIhAAAARI "-" /apache/20230402/20230402-2148/20230402-214813-ZCnb_YjvQ81K0fl0xJUIhAAAARI 0 2935 md5:279ba58f31cb12da9b4b07c85ff05bda 
modsec_audit.log:4484:www.autojanser.ch 148.72.232.65 - - [02/Apr/2023:21:57:43.765826 +0200] "POST /xmlrpc.php HTTP/1.1" 403 199 "-" "-" ZCneN4jvQ81K0fl0xJUI5AAAARE "-" /apache/20230402/20230402-2157/20230402-215743-ZCneN4jvQ81K0fl0xJUI5AAAARE 0 2914 md5:95a2b59dfe8154d07f45f933e9dbef8b 
```

What [Definition] do I need to make sure fail2ban can easily ban on server level the IPs from blacklist when they return.

I tried ```Access denied with code 403 (phase 1). Matched phrase "/HOST/"``` from the modsec_debug.log, also what is best log to use in the rule.

Any help is appreciated.

The rule until now looks:

```bash
[modsec]
ports=80,443
logpath=/var/log/httpd/modsec_audit.log tail
banaction=firewallcmd-ipset-reset
enabled=true
bantime=600
findtime=300
maxretry=1
filter=modsecurity-blacklist
chain=INPUT
```

Yet the filter modsecurity-blacklist does not seem to make it working to match the IP and ban them.

**Comment by sebres** [comment, marked as answer]
The entries from modsec_audit.log look like normal access-log, so:
```ini
failregex = ^\S+ <ADDR> \S+ \S+ (?:\[\] )?\"[A-Z]+ /[^\"]+" 403\s
```

By the way, you don't need a filter for the jail, one can add `filter =` (to empty value) and set `failregex` directly in jail.

**Reply by uhlhosting** [reply]
Hi @sebres thanks so much, so I found that it works fine for the 403 rules from modsec_audit, yet if you look at the above log, I would like to be able to block also the modsec_debug entries in the form:

```bash
modsec_debug.log:1199:[02/Apr/2023:21:48:13.663397 +0200] [xn--vakuumwrmedmmung-1nbe.ch/sid#55cdb76b4978][rid#7f1dd8004c20][/xmlrpc.php][1] Access denied with code 403 (phase 1). Matched phrase "/148.72.232.65/" at TX:ip_blacklist. [file "/etc/httpd/modsecurity.d/activated_rules/ip-blacklist.conf"] [line "2"] [id "2200000"] [msg "SLR: Client IP in Blacklist."] [tag "AUTOMATION/MALICIOUS"]
```

Status for the jail: modsec
|- Filter
|  |- Currently failed:	0
|  |- Total failed:	15
|  `- File list:	/var/log/httpd/modsec_audit.log
`- Actions
   |- Currently banned:	15
   |- Total banned:	15

**Reply by sebres** [reply]
Why? That are the same attempts (just logged in debug- and in audit-log at the same time, if you'd look at the timestamp).
The problem is - there is no IP in the message, the part of message like `Matched phrase "/148.72.232.65/" at TX:ip_blacklist` may be something different (another matching rule, subnet, reverse-dns or whatever else), so it'd be impossible to identify the IP to ban.
However if you need it nevertheless, here is the RE:
```ini
failregex = ^(?:\[\])?\s*(?:\[[^\]]+\])+ Access denied with code \d+ \(phase \d+\). Matched phrase "/?<ADDR>\b
```
(just if I were you, I wouldn't use that debug-log for monitoring)

**Reply by uhlhosting** [reply]
@sebres I agree, the only issue is that I have a list of like 2000+ ips that have highest rate of attacks, I have the above mentioned modsecurity rule to block the ips based on the list, but this still allow the attacker to hit the server and receive a 403, since is not blocked on firewall level.

**Reply by sebres** [reply]
What has this to do with the fact that the same attempt is logged twice in 2 different logs?
The modsec logs the message in debug-log, then in audit-log then returns 403 to the client.
No matter which log (audit-log or debug-log) fail2ban would monitor, it'd ban the IP only if it finds `maxretry` attempts during `findtime`.
Again, the log where fail2ban find them is completely irrelevant...
Also read FAQ in [wiki :: How fail2ban works](https://github.com/fail2ban/fail2ban/wiki/How-fail2ban-works) if you mean something doesn't get banned or found.

**Reply by uhlhosting** [reply]
@sebres thanks for follow up:

So I have this:

```bash
[modsec]

ports     = 80,443
logpath   = /var/log/httpd/modsec_audit.log tail
banaction = firewallcmd-ipset-reset
enabled   = true
filter    =
#datepattern = ^%%d/%%b/%%Y:%%H:%%M:%%S
bantime   = 9w
findtime  = 3d
maxretry  = 1
failregex = ^\S+ <ADDR> \S+ \S+ (?:\[\] )?\"[A-Z]+ /[^\"]+" 403\s
chain     = INPUT
action    = %(action_)s
            %(action_abuseipdb)s[abuseipdb_category="18,4"]
```


When I comment # the datepattern I get on running `fail2ban-client status modsec` 

```bash
[root@cp fail2ban]# fail2ban-server status modsec
Status for the jail: modsec
|- Filter
|  |- Currently failed:	0
|  |- Total failed:	23
|  `- File list:	/var/log/httpd/modsec_audit.log
`- Actions
   |- Currently banned:	116
   |- Total banned:	116
   `- Banned IP list:	xxx.xxx.xxx.xxx...
```

Any ideas why? And how to make sure that there are not failed, while in same time they are getting blocked?

When I uncomment the #datepattern then I do not have failed entries, yet I have a feeling it stops banning them:

```bash
[root@cp fail2ban]# fail2ban-server status modsec
Status for the jail: modsec
|- Filter
|  |- Currently failed:	0
|  |- Total failed:	0
|  `- File list:	/var/log/httpd/modsec_audit.log
`- Actions
   |- Currently banned:	136
   |- Total banned:	136
   `- Banned IP list:	xxx.x.x.x...
```

**Reply by uhlhosting** [reply]
This is indeed very weird, so when I disable the `datepattern` its keeps the last blocks from memory/db yet does not ban new ips/ old ips even when I increased the `findtime`

With the datepattern on, it bans yet I also get on failed entries, without it stops banning anymore ips.

```bash
[root@cp fail2ban]# fail2ban-server status modsec
Status for the jail: modsec
|- Filter
|  |- Currently failed:	0
|  |- Total failed:	13
|  `- File list:	/var/log/httpd/modsec_audit.log.1 /var/log/httpd/modsec_audit.log.4.gz /var/log/httpd/modsec_audit.log.3.gz /var/log/httpd/modsec_audit.log.2.gz /var/log/httpd/modsec_audit.log
`- Actions
   |- Currently banned:	149
   |- Total banned:	149
   `- Banned IP list:	
[root@cp fail2ban]# 
```

**Comment by uhlhosting** [comment]
Also noticed this:

```bash
['start', 'modsec']]] has failed. Received RegexException("Unable to compile regular expression '[client (?:\\[?(?:(?:::f{4,6}:)?(?P<ip4>(?:\\d{1,3}\\.){3}\\d{1,3})|(?P<ip6>(?:[0-9a-fA-F]{1,4}::?|::){1,7}(?:[0-9a-fA-F]{1,4}|(?<=:):)))\\]?|(?P<dns>[\\w\\-.^_]*\\w))] ModSecurity: Access denied with code 406.+BAD BOT':\nunbalanced parenthesis at position 97",)
```

and it seems when I run sudo zgrep 'Ban' /var/log/fail2ban.log* | grep modsec I do not see the module loading but only this message above thats the last part, the ERROR message is way bigger.

**Reply by sebres** [reply]
> And how to make sure that there are not failed, while in same time they are getting blocked?

`Currently failed` (0) means the failures that are found right now (but not yet banned), in your case no one.
`Total failed` (23) means - are the failures found since last restart of fail2ban or jail.

I don't understand what you want.

> When I uncomment the #datepattern then I do not have failed entries

The datepattern `^%%d/%%b/%%Y:%%H:%%M:%%S` is not valid - `^` is a start-anchor, it means that timestamp takes place at begin of string, in your audit.log it takes place in the middle of a message line within square brackets.

>  its keeps the last blocks from memory/db yet does not ban new ips/ old ips even when I increased the findtime

Fail2ban knows the last position in log (stored in DB), and would never rescan the log... no matter how large is your findtime (it'd only work for new entries in log, but with `maxretry = 1` is doesn't matter at all, because IP gets banned immediately after first attempt).
Here is more about - https://github.com/fail2ban/fail2ban/issues/2909#issuecomment-758036512

**Comment by uhlhosting** [comment]
@sebres thats verified that the date pattern is wrong, yet when I go against the fail2ban-regex, I get:

I have moved the regexfilter inside filter.d to easier debug it.

```bash
[root@cp filter.d]# fail2ban-regex /var/log/httpd/modsec_audit.log /etc/fail2ban/filter.d/modsecurity.conf

Running tests
=============

Use   failregex filter file : modsecurity, basedir: /etc/fail2ban
Use      datepattern : {^LN-BEG} : Default Detectors
Use         log file : /var/log/httpd/modsec_audit.log
Use         encoding : UTF-8


Results
=======

Failregex: 0 total

Ignoreregex: 0 total

Date template hits:

Lines: 5350 lines, 0 ignored, 0 matched, 5350 missed
[processed in 0.41 sec]

Missed line(s): too many to print.  Use --print-all-missed to print all 5350 lines
```

when I print all lines I get:

```bash
|  www.darkbot.in 190.107.177.235 - - [03/Apr/2023:20:47:16.298516 +0200] "POST /xmlrpc.php HTTP/1.1" 403 199 "-" "-" ZCsfNHd5Pve9Phd1CuUuPgAAAQY "-" /apache/20230403/20230403-2047/20230403-204716-ZCsfNHd5Pve9Phd1CuUuPgAAAQY 0 2751 md5:d5d9956c82e39e130114b9d673159d01
|  www.autojanser.ch 85.95.231.40 - - [03/Apr/2023:20:48:27.777147 +0200] "POST /xmlrpc.php HTTP/1.1" 403 199 "-" "-" ZCsfe0Ann7ms3xU4zmcnGwAAAIE "-" /apache/20230403/20230403-2048/20230403-204827-ZCsfe0Ann7ms3xU4zmcnGwAAAIE 0 2180 md5:74563171f40400196f45da149c27f2f3
|  xn--vakuumwrmedmmung-1nbe.ch 34.28.16.79 - - [03/Apr/2023:20:48:46.714618 +0200] "POST /xmlrpc.php HTTP/1.1" 403 199 "-" "-" ZCsfjnd5Pve9Phd1CuUuUwAAAQc "-" /apache/20230403/20230403-2048/20230403-204846-ZCsfjnd5Pve9Phd1CuUuUwAAAQc 0 2188 md5:f7829e08f6371f5a4dd6e8a5b2056cf2
```

I just want to make sure that the rule works as designed and no issues are the config.
Would had expected to get something like **5350 matched, 0 missed**


When I look at present rule that works just fine:

Use   failregex filter file : modsecurity-malware, basedir: /etc/fail2ban
Use      datepattern : %d/%b/%Y:%H:%M:%S : Day/MON/Year:24hour:Minute:Second
Use         log file : /var/log/httpd/modsec_audit.log
Use         encoding : UTF-8


Results
=======

**Failregex: 77 total**
|-  #) [# of hits] regular expression
|   1) [77] ^\S+\s*<HOST>\s+(?:\S+\s+){2}\[[^]]+\]\s+"[^\"]+"\s+406\s
`-

Ignoreregex: 0 total

Date template hits:
|- [# of hits] date format
|  [5447] Day/MON/Year:24hour:Minute:Second
`-

Lines: 5447 lines, 0 ignored, **77 matched,** 5370 missed


I get the lines 77 matched . Not the case of this current rule...

**Reply by sebres** [reply]
Do you see that?

> `Use      datepattern : {^LN-BEG} : Default Detectors`

This indicate that datepattern of filter is set to `{^LN-BEG}` (this means all patterns fail2ban supports, but also anchored to the line begin). 
Do you use some include in your filter? For instance `common.conf` or maintainer includes may have this by default.

You don't need that include here, so just remove it... As already said failregex is completely enough.

Or set the datepattern to empty value or to `\s\[{DATE}\]`.

**Reply by uhlhosting** [reply]
Now I added:

```bash
datepattern = \s\[{DATE}\]
failregex = ^\S+ <ADDR> \S+ \S+ (?:\[\] )?\"[A-Z]+ /[^\"]+" 403\s
```

Will try this and also with empty value. Yes I had common.conf included I was thinking it helps. Removed it now.

**Reply by uhlhosting** [reply]
My complete rule now looks like:

```
ports     = 80,443
logpath   = /var/log/httpd/modsec_audit.log tail
banaction = firewallcmd-ipset-reset
enabled   = true
filter    =
bantime   = 52w
findtime  = 1d
maxretry  = 2
datepattern = \s\[{DATE}\]
failregex = ^\S+ <ADDR> \S+ \S+ (?:\[\] )?\"[A-Z]+ /[^\"]+" 403\s
chain     = INPUT
action    = %(action_)s
            %(action_abuseipdb)s[abuseipdb_category="18,4"]
```

**Reply by uhlhosting** [reply]
The only thing is , not sure how to use the `fail2ban-regex`, to check I have tried to include the regex, and did not worked, that is why I have tried with the `filter.d/file.conf` , to use it as `fail2ban-regex /var/log/httpd/modsec_audit.log /etc/fail2ban/filter.d/modsecurity.conf` I tried something like:  `fail2ban-regex /var/log/httpd/modsec_audit.log "^\S+ <ADDR> \S+ \S+ (?:\[\] )?\"[A-Z]+ /[^\"]+"` yet did not worked.

**Reply by sebres** [reply]
I already wrote how filter may look.
Either remove from the filter the includes (unneeded here) or set unanchored date pattern (or empty one).
As for regex in command line, enclose it in single quotes instead of double quotes (they have special escape sequences, you can read about it in man of bash or Linux shell)

**Reply by uhlhosting** [reply]
@sebres thanks for the time and patience, solved the issues, and now seem to be working all well.
